Detect duplicate metric names

bassamadnan opened the pull request:
Added logic for detecting duplicate metric names, this includes handling different queries based on the same version, validating before processing the file and maintaining a global (`configuration :: metric_name_trie`) list of metrics processed.

Also with this PR we can only read either the YAML/JSON files, or the internal metrics (not both together). Also changed `pg_db_conn` to `pg_db_conn_ssl` to resolve the name conflict.

jesperpedersen commented:
@bassamadnan Can't we use ART instead of trie ?
@Jubilee101 Cc:

bassamadnan commented:
> @bassamadnan Can't we use ART instead of trie ? @Jubilee101 Cc:

I suppose we could, but i have never worked with ART before, and since this is just being done during startup i wanted to keep things simple (and easier to debug).

resyfer commented:
@bassamadnan You can try looking at `prometheus_client.c` (as already mentioned above) and even [bridge.c](http://github.com/pgexporter/pgexporter/blob/main/src/libpgexporter/bridge.c#L965) for an example of the usage of the ART API. You really don't need to know the internals once you've figured out what components the API needs or how the API is used. In other words that's the reason why the API exists :D

bassamadnan commented:
Fortunately the code was very similliar, so I could just take a refference from existing APIs and port it. I have removed trie and just use existing ART.

> https://db.in.tum.de/~leis/papers/ART.pdf

Thanks for this, was a fun read, still abit complicated i'd say and im not sure if i completely understand it. But the variables and structs do make much more sense now.



> 
> I don't understand the concern here, can you elaborate? `MAX_PREFIX_LEN` is the size of prefix "cache", we don't save the rest of the prefix if it's longer than that. Having the first 10 chars of a prefix serves as a filter for those whose prefixes diverge before the first 10 characters. If you use trie, which doesn't do path compression at all, it costs you even more space because each char in the prefix is now a trie node with additional memory overhead.

I was  confused with  what `MAX_PREFIX_LEN` meant, reading the paper, watching few videos and seeing the code did clarify it.

> And I'm sure you can still do that. You can also look into regex if you want to do faster validation.

Yup,  just added a simple function in `utils`



> @bassamadnan You can try looking at `prometheus_client.c` (as already mentioned above) and even [bridge.c](https://github.com/pgexporter/pgexporter/blob/main/src/libpgexporter/bridge.c#L965) for an example of the usage of the ART API. You really don't need to know the internals once you've figured out what components the API needs or how the API is used. In other words that's the reason why the API exists :D

Yes. This was very helpful. Thanks!

PTAL.

jesperpedersen commented:
Merged.

Thanks for your contribution !
